### DISCUSSION TITLE
Slow Firestore Queries in Flutter Web After Version Upgrade

### DISCUSSION THREAD

**Discussion by Ultranmus** [firebase/flutterfire#12048] [Q&A]
**Description:**

I'm currently facing a performance issue with Firestore queries in my Flutter web application after upgrading from version 3.13.7 to the latest one. Previously, queries fetching 12 documents from a Firestore collection, totaling around 72.65 KB of data, took 1-2 seconds. However, post-upgrade, the same queries are now taking around 30 seconds, leading to server timeouts.

**Here are some key details:**

This issue is consistent across both the production and test environments. I've confirmed that the problem started after the Flutter version upgrade. Notably, any release of the application before the version upgrade works normally, with queries taking 1-2 seconds. Even when using the exact production release from my Git repository, the slow query issue persists.

```
void fetchData() async {
  
  var query = FirebaseFirestore.instance
      .collection('articles')
      .orderBy('pub_date', descending: true)
      .limit(12);

  print("Before Firestore Query: ${DateTime.now()}");

  var result = await query.get();

  print("After Firestore Query: ${DateTime.now()}");

}

```

**Additional Information:**

- **Flutter Version Before Upgrade:** 3.13.7
- **Flutter Version After Upgrade:** Latest version
- **Cloud Firestore SDK Version:** 4.12.2
- **It's worth noting that there is no problem with Firestore, as the Android and iOS apps for it are working perfectly fine.**

**Attempted Solutions:** 

- I tried downgrading the Flutter version back to 3.13.7, but it did not resolve the issue.
- I tried the latest cloud_firestore version also but did not worked out.
- I've isolated this problem in a minimal environment, and the delay is consistent across the demo project, indicating that it's not specific to the application's complexity.
- I cloned a previous version of the code that was working fine in production. After rebuilding Flutter web with the previously working code, the performance issue persisted.

I'm seeking guidance on where the problem might be originating—whether it's in my Flutter web application, Firestore, or somewhere else. Any suggestions or insights on how to troubleshoot and resolve this performance issue would be greatly appreciated.

**Comment by Ehesp** [comment]
This sounds odd for sure. When exactly are you triggering the query? Is it when your application is starting up or on an explicit event?

Wondering whether the main js thread is being overloaded or something

**Comment by Ultranmus** [comment]
@Ehesp well I am calling it in my first screen using a stream builder. 
The odd thing that I have not changed even a single bit of code but the query are slow whereas the previous production is working perfectly.

**Comment by Abdaallaa** [comment]
i have the same issue week ago

**Comment by Ultranmus** [comment]
I have also tested firestore in react with same firestore project and the prblm persists. Where else in Android and ios, no latency. I also saw similar issue in react
https://github.com/firebase/firebase-js-sdk/issues/7860
The prblm is in new firebase js sdk, I think the prblm in flutter also generate from there.

**Reply by Abdaallaa** [reply]
Unfortunately, downgrading the Flutter and Firestore versions is not a solution. However, Firestore for mobile works normally.

**Comment by Ultranmus** [comment]
@Abdaallaa same for me, I also tried downgrading everything from flutter to each plugin and package version but nothing worked. But for Android and ios it works normally.

**Comment by skiluk** [comment]
Is there any update on a fix for this? We are experiencing this same issue.

**Comment by Ultranmus** [comment]
@skiluk I couldn't find a solution, so I opted to discard the Firestore package and utilize Google Firestore APIs instead.
https://firebase.google.com/docs/firestore/reference/rpc
